storage,batcheval: add option to use fast sst rewriter

dt opened the pull request:
Release note: none.

dt committed:
storage,batcheval: add option to use fast sst rewriter

Release note: none.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/76074)
<!-- Reviewable:end -->

erikgrinaker commented on pkg/storage/sst.go:
```diff
@@ -224,14 +225,23 @@ func CheckSSTConflicts(
 	sstOut := &MemFile{}
 	sstOut.Buffer.Grow(len(sst))
+	if concurrency > 0 && !from.IsEmpty() {
+		from := EncodeMVCCKeyToBuf([]byte{}, MVCCKey{Timestamp: from})[mvccEncodedTimeSentinelLen:]
```
These can use `encodeMVCCTimestampSuffix()`.

erikgrinaker commented on pkg/kv/kvserver/batcheval/cmd_add_sstable.go:
This is going to fail for clients that don't set `SSTTimestamp`, but that's ok for now since it's not long for this world anyway.

dt replied:
Is it? I thought the IsEmpty condition would handle that?

erikgrinaker replied:
You're right, missed that.

erikgrinaker reviewed (approved):
🏎

erikgrinaker commented on pkg/kv/kvserver/batcheval/cmd_add_sstable.go:
```diff
@@ -37,6 +38,14 @@ func init() {
 	RegisterReadWriteCommand(roachpb.AddSSTable, DefaultDeclareIsolatedKeys, EvalAddSSTable)
 }
 
+var addSSTableRewriteConcurrency = settings.RegisterIntSetting(
```
Can we make this metamorphic?

dt replied:
I was going to leave it as zero for now until I do some more production tests?

dt replied:
added

dt commented:
TFTR!

bors r+

dt commented:
bors r-

missed the suffix comment

dt commented:
Updated. Added a metamorphic bool to take the rewrite path some of the time even if the timestamp wasn't pushed, to ensure we get some coverage of it.

erikgrinaker reviewed (approved):
LGTM, thanks!

dt commented:
TFTR!

bors r+
